core/src/appleMain/kotlin/CBPeripheralCoreBluetoothPeripheral.kt

```diff
@@ -393,9 +402,47 @@ internal class CBPeripheralCoreBluetoothPeripheral(
         }
 
         val platformDescriptor = discoveredServices.obtain(descriptor)
-        return connection.execute<DidUpdateValueForDescriptor> {
+        val updatedDescriptor = connection.execute<DidUpdateValueForDescriptor> {
             centralManager.read(cbPeripheral, platformDescriptor)
-        }.descriptor.value as NSData
+        }.descriptor
+
+        // CBDescriptor "value" property type depends on cbDescriptor.UUID.UUIDString
+        // https://developer.apple.com/documentation/corebluetooth/characteristic-descriptors
+        // see type conversion table https://github.com/JuulLabs/kable/pull/706
+        when (updatedDescriptor.UUID.UUIDString) {
+            CBUUIDCharacteristicFormatString -> {
+                return updatedDescriptor.value as NSData
+            }
+            CBUUIDCharacteristicUserDescriptionString,
+            -> {
+                return (updatedDescriptor.value as String)
+                    .encodeToByteArray()
+                    .toNSData()
+            }
+            CBUUIDCharacteristicExtendedPropertiesString,
+            CBUUIDClientCharacteristicConfigurationString,
+            CBUUIDServerCharacteristicConfigurationString,
+            -> {
+                return (updatedDescriptor.value as NSNumber)
+                    .unsignedShortValue
+                    .toNSData()
+            }
+            CBUUIDL2CAPPSMCharacteristicString,
+            -> {
+                return (updatedDescriptor.value as UInt16).toNSData()
+            }
+            CBUUIDCharacteristicAggregateFormatString,
+            -> {
+                logger.warn {
+                    message = "Best effort descriptor value conversion for undocumented $CBUUIDCharacteristicAggregateFormatString uuid"
+                }
+                return (updatedDescriptor.value as? NSData) ?: byteArrayOf().toNSData()
+            }
+            else -> {
+                logger.warn { message = "cannot read descriptor for unknown uuid string ${updatedDescriptor.UUID.UUIDString}" }
+                return byteArrayOf().toNSData()
+            }
+        }
```

Comment:
As I was looking how others read descriptors, I stumbled on [this](https://github.com/vokal/flutter_blue/blob/3b6f1552d87554f47580449b16e8f3252b1832ff/ios/Classes/FlutterBluePlugin.m#L555-L565):

```objc
// TODO: Convert NSNumber types too
if( [descriptor.value isKindOfClass:[NSData class]] ) {
    return descriptor.value;
}
if( [descriptor.value isKindOfClass:[NSString class]] ) {
    return [descriptor.value dataUsingEncoding:NSUTF8StringEncoding];
}
if( [descriptor.value isKindOfClass:[NSNumber class]] ) {
    int value = [descriptor.value intValue];
    return [NSData dataWithBytes:&value length:sizeof(value)];
}
```

Which hints that the data we get from Core Bluetooth carries the type info, meaning we could possibly do something similar to the code snippet above?

```suggestion
        return when (val value = updatedDescriptor.value) {
            is NSData -> value
            is NSString -> value.dataUsingEncoding(NSUTF8StringEncoding)
                ?: byteArrayOf().toNSData().also {
                    logger.warn {
                        message = "Failed to decode descriptor"
                        detail("type", "NSString")
                        detail(descriptor)
                    }
                }
            is NSNumber -> memScoped {
                NSData.dataWithBytes(
                    bytes = alloc<IntVar>().also { it.value = value.intValue }.ptr,
                    length = sizeOf<IntVar>().convert(),
                )
            }
            else -> byteArrayOf().toNSData().also {
                logger.warn {
                    message = "Unable to read descriptor of unknown type"
                    detail(descriptor)
                }
            }
```

_Though, I'm not sure what is meant by the `// TODO: Convert NSNumber types too` code comment._

Reply:
Additional instances where the type of the descriptor value was used to process it:
- [FlutterBluePlusPlugin.m#L2137-L2160](https://github.com/boskokg/flutter_blue_plus/blob/0d439a98d5b5ce76ce51a95c9ca93e03e04c1e8a/ios/Classes/FlutterBluePlusPlugin.m#L2137-L2160)
- [BleManager.m#L121-L142](https://github.com/AndrewDongminYoo/react-native-ble-manager/blob/80634af6ac4cda7e203fd936812c2efe3820ef1e/ios/BleManager.m#L121-L142)

Reply:
Yes this approach of checking types directly can work as well.
It was actually my first implementation, but then I opted on checking the `descriptor.UUID.UUIDString` as it was what apple doc suggested and It probably felt less _arbitrary_ to somebody looking at code for the first time.

> Though, I'm not sure what is meant by the // TODO: Convert NSNumber types too code comment.

I guess this is a drawback of this approach.
I think that what the comment is trying to say is that `NSNumber` is always being converted to an `Int` (i.e. 4 bytes), but it is actually a class that can contain any type of number, such as `Long` (8 bytes), `Short` (2 bytes) etc.
This means that if `NSNumber` actually contains a `Long`, then we lost 4 bytes of information by converting to `Int`. 

This approach is also used on [Dotnet](https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/blob/master/Source/Plugin.BLE/Apple/Descriptor.cs) which  tries to workaround this issue by casting to `UInt64` (the biggest possible number type, so no bytes are lost).

But even with or without Dotnet workaround, there a still probably some inconsistencies across platforms.

For example take `CBUUIDClientCharacteristicConfigurationString`.
The Bluetooth specification states:

> The Attribute Value shall be two octets in length and shall be set to the characteristic descriptor value

But with this approach we will return  either 4 bytes (without Dotnet workaround) or 8 bytes (with Donet workaround), when the specification states that should be 2 bytes.
Furthermore, on other platforms (i.e. Android, JS) 2 bytes are returned (I didn't test, it's just my guess, but seems very likely)

I'm not trying to dismiss this approach as it is indeed simpler, less verbose and seems to be widely adopted by other libraries / Multiplatform frameworks, but just wanted to highlight some of the potentials limitations.

Reply:
Thanks for the insight! I'm not attached to one approach over the other. As you've identified, they both have trade offs.

Perhaps a mix of the two approaches might make the most sense? Where we first process based on type, then when an `NSNumber`: determine what size number to process as based on UUID.

Using only UUID, as you currently have it, does seem like a really good approach (and it's nice to follow recommendations set forth by the documentation) but being that the documentation leaves a lot to be desired (at times it isn't clear exactly what type will be provided), it feels safer to process by type first (for the non-number types). 🤷‍♂️

Reply:
> Perhaps a mix of the two approaches might make the most sense? Where we first process based on type, then when an NSNumber: determine what size number to process as based on UUID.

Yes I guess we could go along this way since we couldn't find anywhere the expected type returned by Apple.

I've updated the PR by reworking the conversion approach:

- For types that can be converted directly to NSData without ambiguity (i.e. NSString, NSData, UInt16), I convert it directly without further checks on UUID string
- For NSNumber (that can't be converted directly to NSData without ambiguity):
  -  I convert it to the documented numeric type by checking on the UUID string where possible
  - otherwise I convert it to `ULong` to avoid information loss (Dotnet approach)

Let me know your feedback!